default/python/AI/PlanetUtilsAI.py

```diff
@@ -166,3 +163,40 @@ def get_systems(planet_ids: Sequence[PlanetId]) -> Sequence[SystemId]:
     # TODO discuss change return type to set
     universe = fo.getUniverse()
     return [universe.getPlanet(pid).systemID for pid in planet_ids]
+
+
+def get_planet_opinion(feature: str) -> (Set[PlanetId], Set[PlanetId], Set[PlanetId]):
+    """
+    Returns sets of empire planets that like, are neutral and dislike the given feature
+    """
+    # default: feature not in any like or dislike set, all neutral
+    default = (set(), set(get_owned_planets_by_empire()), set())
+    return _calculate_get_planet_opinions().get(feature, default)
+
+
+@cache_for_current_turn
+def _calculate_get_planet_opinions():  # -> Dict[str, (Set[PlanetId], Set[PlanetId], Set[PlanetId])]:
```

Comment:
I couldn't make the typing work, but for a private function it is not that important anyway.

Reply:
`Union`?

Reply:
```suggestion
def _calculate_get_planet_opinions(): -> Dict[str, Tuple[Set[PlanetId], Set[PlanetId], Set[PlanetId]]]``` 
 ```

Reply:
Nut actually we could replace tuple with named tuple.  It's lightweight as a tuple but convenient as a class. 

So instead of accession attributes by index, you could do it with name.

Python 3.6 has a very convenient way to declare it. 
https://docs.python.org/3/library/typing.html#typing.NamedTuple

```python
from typing import NamedTuple, Set

from fo_typing import PlanetId


class Preferences(NamedTuple):
    likes: Set[PlanetId]
    neutral: Set[PlanetId]
    dislikes: Set[PlanetId]
```

Reply:
Thank you for the hint, that definitely better.